java/arcs/sdk/Encoding.kt

```diff
@@ -0,0 +1,165 @@
+
+    fun done(): Boolean = bytes.isEmpty() || bytes[0] == 0.toByte()
+
+    fun upTo(sep: Char): ByteArray {
+        val ind = bytes.indexOf(sep.toByte())
+        if (ind == -1) {
+            error("Packaged entity decoding failed in upTo()\n")
+        }
+        val chunk = bytes.sliceArray(0..(ind - 1))
+        bytes = bytes.sliceArray((ind + 1)..(bytes.size - 1))
+        return chunk
+    }
+
+    fun getInt(sep: Char): Int = upTo(sep).utf8ToString().toInt()
+
+    fun chomp(len: Int): ByteArray {
+        // TODO: detect overrun
+        val chunk = bytes.sliceArray(0..(len - 1))
+        bytes = bytes.sliceArray(len..(bytes.size - 1))
+        return chunk
+    }
+
+    fun validate(token: String) {
+        if (chomp(token.length).utf8ToString() != token) {
+            throw IllegalArgumentException("Packaged entity decoding failed in validate()")
+        }
+    }
+
+    fun decodeText(): String = chomp(getInt(':')).utf8ToString()
+
+    fun decodeNum(): Double = upTo(':').utf8ToString().toDouble()
+
+    fun decodeBool(): Boolean = chomp(1).utf8ToString() == "1"
+
+    companion object {
+        fun decodeDictionary(bytes: ByteArray): Map<String, String> {
+            val decoder = StringDecoder(bytes)
+            val dict = mutableMapOf<String, String>()
+
+            var num = decoder.getInt(':')
+            while (num-- > 0) {
+                val klen = decoder.getInt(':')
+                val key = decoder.chomp(klen).utf8ToString()
+
+                val vlen = decoder.getInt(':')
+                val value = decoder.chomp(vlen).utf8ToString()
+
+                dict[key] = value
+            }
+
+            return dict
+        }
+    }
+}
+
+class StringEncoder(
+    private val buffers: MutableList<ByteArray> = mutableListOf(),
+    private var size: Int = 0
+) {
+    fun encodeDictionary(dict: Map<String, Any?>): StringEncoder {
+        addStr("${dict.size}:")
+        for ((key, value) in dict) {
+            addBytes("", key.stringToUtf8())
+            encodeValue(value)
+        }
+        return this
+    }
+
+    fun encodeList(list: List<Any>): StringEncoder {
+        addStr("${list.size}:")
+        list.forEach { encodeValue(it) }
+        return this
+    }
+
+    fun encodeValue(value: Any?) {
+        when (value) {
+            is String -> addBytes("T", value.stringToUtf8())
+            is Boolean -> addStr("B${if (value) 1 else 0}")
+            is Double -> addStr("N$value:")
+            is Map<*, *> -> {
+                @Suppress("UNCHECKED_CAST")
+                val se = StringEncoder().encodeDictionary(value as Map<String, Any?>)
+                addBytes("D", se.toByteArray())
+            }
+            is List<*> -> {
+                @Suppress("UNCHECKED_CAST")
+                val se = StringEncoder().encodeList(value as List<Any>)
+                addBytes("A", se.toByteArray())
+            }
+            else -> throw IllegalArgumentException("Unknown expression.")
```

Comment:
"Unknown value" or "Unsupported value type" might be more accurate?

Reply:
Done.